### ISSUE TITLE
Provide support for configuring client auth.

### ISSUE THREAD

**Issue by larkinkevin** [projectcontour/contour#433]
- Envoy supports client auth.  It would be great if Contour did too (ie. if Contour exposed this particular piece of Envoy config).
- Contour already supports server auth (via Ingress' "tls" stanza).  Adding support for client auth would unblock users who want to use Contour in an mTLS setup.

**Comment by ihrwein** [comment]
I wanted to raise a similar issue then managed to convince myself that although this feature would be useful it's not aligned with this project's goals (note that I'm just somebody from the Internet who found this project some days ago.).

My reasoning is that different teams can have different authentication requirements. Contour is to have one Ingress that delegates the traffic to team-managed services. Those services can be simple services or API gateways that hides other services and those API gateways would be a better place for authentication. Just my two cents.

**Comment by rosskukulinski** [comment]
@davecheney this came up with a customer recently - will likely be a blocking issue in Q1 or Q2 of next year.  

Relevant nginx annotation is here: https://github.com/kubernetes/ingress-nginx/blob/master/docs/user-guide/nginx-configuration/annotations.md#client-certificate-authentication

Envoy docs: https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/ssl#enabling-certificate-verification

**Comment by davecheney** [comment]
Seems reasonable to me. Let’s pencil this in for 0.9 for jan/feb 2019

**Comment by hlarubi** [comment]
Hi,

Is this item still on the roadmap for 1Q2019?

Thanks!

**Comment by davecheney** [comment]
Due to the need to ship Contour 0.11 to address the security issue in Envoy 1.9.0 this issue has been moved to the 0.12 milestone.

@hlarubi it is likely that this issue will not make the 0.12 milestone either. @stevesloka made a start on this in #978 but given there is no design document for this yet, it's unlikely this will be completed in the 0.12 milestone. I suspect that implementing this feature will have to wait for SDS support to land, #898

**Comment by uablrek** [comment]
Relevant Envoy doc; https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/cert.proto.html?highlight=require_client_certificate

Example;
```
      tls_context:
        require_client_certificate: true
        common_tls_context:
          validation_context:
            verify_certificate_spki: j1fFFSBuF7S/Hzk5xsIq7N7wvVm8xdMpMnt0OUsz5rs=
```

**NOTE**; `require_client_certificate: true` will be silently ignored unless a `validation_context` is specified! See; https://github.com/envoyproxy/envoy/issues/3887

In the example Subject Public Key Information (SPKI) is used to validate the client certificate. As I understand this does not have to be secret but is generated outside the cluster without any risk of exposing the private key.

The spki is generated as described in the envoy doc with;
```
openssl x509 -in default/cert/kahttp.crt -noout -pubkey \
  | openssl pkey -pubin -outform DER \
  | openssl dgst -sha256 -binary \
  | openssl enc -base64
```

**Comment by davecheney** [comment]
Thank you. Can you please raise a new issue if this is a bug. Please note
that validation is _not_ client side validation, its simply validation of
the properties of the remove service.

Nevertheless, please raise a new issue, and we'll talk about it.

**Comment by uablrek** [comment]
I think certain backends will want client certificate validation so IMO this should be specified as a [Contour specific Service annotation](https://github.com/heptio/contour/blob/master/docs/annotations.md#contour-specific-service-annotations).

Example;
```
contour.heptio.com/client-certificate-validate.spki=j1fFFSBuF7S/Hzk5xsIq7N7wvVm8xdMpMnt0OUsz5rs=
```

**Comment by davecheney** [comment]
I agree. Client certification validation is useful over and above what we shipped in 0.12. 

Please file a new issue and we’ll figure out where it goes in the priority order.

**Comment by uablrek** [comment]
@davecheney No this is not a bug. It is info about client certificate validation (mTLS) which I thought was the subject of this enhancement issue(?)

I have learned and tested this on a stand-alone Envoy to see what can be done and how. But the goal is support for client cerificate validation in contour.

This may become a blocker for us and I would be happy to contribute, if nothing else with alpha testing.

**Comment by davecheney** [comment]
You can use an issue for a feature request. That’s totally fine.

**Comment by uablrek** [comment]
I can raise a new issue. Can the title be;
```
Add support for client cerificate validation (mTLS)
```

**Comment by davecheney** [comment]
I don’t know how to say yes any more strongly than I have said it already. Please raise a new issue.

**Comment by davecheney** [comment]
Assigning to 0.15 tentatively. We don’t have anyone available to work on this feature so it may get dropped from the 1.0 release.

**Comment by davecheney** [comment]
Hello,

As described in https://github.com/heptio/contour/issues/433#issuecomment-503014942 the assignment to 0.15 was tentative and we did not have anyone assigned to work on this feature. 01.5, our last development cycle before the 1.0 freeze, closes next week and this feature will not be ready. I am moving this to the unplanned milestone and we will revisit it after Contour 1.0 in November.

**Comment by jpeach** [comment]
Reading back, it seems like this issue is all about TLS client certificate authentication. Since we landed that feature in 1.4.0, I think that this is a duplicate of #1090. We have added the bare essentials of certificate authentication, so I expect that there are features and use cases in this area that are  not yet covered. If you have one, please file a new issues so that we can track adding those specifically.

This duplicates #1090.
